native/cocos/renderer/gfx-base/memory/TLSFPool.cpp

```diff
@@ -0,0 +1,318 @@
+            block = nullptr;
+        }
+    }
+}
+
+uint64_t TLSFPool::roundUpSize(uint64_t size) {
+    uint64_t roundSize = size;
+    if (size >= MIN_BLOCK_SIZE) {
+        roundSize += (1ULL << (bitScanReverse(size) - SECOND_LEVEL_INDEX)) - 1;
+    } else {
+        roundSize += SMALL_BUFFER_STEP - 1;
+    }
+    return roundSize;
+}
+
+void TLSFPool::levelMapping(uint64_t size, uint32_t &fl, uint32_t &sl) {
+    if (size < MIN_BLOCK_SIZE) {
+        fl = 0;
+        sl = static_cast<uint32_t>(size) / SMALL_BUFFER_STEP;
+    } else {
+        fl = bitScanReverse(size);
+        sl = static_cast<uint32_t>(size >> (fl - SECOND_LEVEL_INDEX)) ^ (1U << SECOND_LEVEL_INDEX);
+        fl -= (FIRST_LEVEL_SHIFT - 1);
+    }
+}
+
+std::pair<uint32_t, uint32_t> TLSFPool::levelMapping(uint64_t size) {
+    uint32_t fl = 0;
+    uint32_t sl = 0;
+    levelMapping(size, fl, sl);
+    return {fl, sl};
+}
+
+TLSFPool::Block *TLSFPool::searchDefault(uint64_t size, uint64_t alignment, uint64_t &alignOffset) {
+    // quick check free list
+    if (_freeBlockCount == 0) {
+        return check(*_nullBlock, size, alignment, alignOffset) ? _nullBlock : nullptr;
+    }
+
+    // try next level
+    {
+        uint64_t roundupSize = roundUpSize(size);
+        auto [fl, sl] = levelMapping(roundupSize);
+        auto *freeBlock = searchFreeBlock(fl, sl);
+        while (freeBlock != nullptr) {
+            if (check(*freeBlock, size, alignment, alignOffset)) {
+                return freeBlock;
+            }
+            freeBlock = freeBlock->nextFree;
+        }
+    }
+
+    // try null block again
+    if (check(*_nullBlock, size, alignment, alignOffset)) {
+        return _nullBlock;
+    }
+
+    // try best fit
+    {
+        auto [fl, sl] = levelMapping(size);
+        auto *freeBlock = searchFreeBlock(fl, sl);
+        while (freeBlock != nullptr) {
+            if (check(*freeBlock, size, alignment, alignOffset)) {
+                return freeBlock;
+            }
+            freeBlock = freeBlock->nextFree;
+        }
+    }
+    return nullptr;
+}
+
+TLSFPool::Block *TLSFPool::searchFreeBlock(uint32_t &fl, uint32_t &sl) {
+    uint32_t slMap = _slBitMask[fl] & (~(0U) << sl);
+    if (slMap == 0) {
+        uint32_t flMap = _flBitMask & (~(0U) << (fl + 1));
+        if (flMap == 0) {
+            // no available free blocks
+            return nullptr;
+        }
+
+        fl = bitScan(flMap);
+        slMap = _slBitMask[fl];
+    }
+    sl = bitScan(slMap);
+    return _freeList[fl][sl];
+}
+
+TLSFPool::Block *TLSFPool::allocate(uint64_t size, uint64_t alignment) {
+    uint64_t realAlignment = std::max(alignment, static_cast<uint64_t>(SMALL_BUFFER_STEP));
+    uint64_t allocSize = utils::align(size, realAlignment);
+
+    uint64_t alignOffset = 0;
+    Block *block = searchDefault(allocSize, realAlignment, alignOffset);
+    if (block != nullptr) {
+        allocateFromBlock(*block, allocSize, alignOffset);
+    }
+    return block;
```

Comment:
Just return it if block is null?

Reply:
Yea, return null if out of memory. The Allocator can check the return value and create a new pool.